File: 백준/Gold/20040. 사이클 게임/사이클 게임.cc
#include <bits/stdc++.h>
using namespace std;
int v, e;
int parent[500001];
int get_parent(int p[], int x){
    if (p[x] != x) {
        p[x] = get_parent(p, p[x]);
    }
    return p[x];
}
void union_parent(int p[], int a, int b){
    a = get_parent(p, a);
    b = get_parent(p, b);
    if (a > b)
        p[a] = b;
    else if (a < b)
        p[b] = a;
    else
        return;
}
int main() {
    ios_base::sync_with_stdio(0);
    cin.tie(0); cout.tie(0);
    int a, b;
    cin >> v >> e;
    for (int i = 0; i < v; i++){
        parent[i] = i;
    }
    for (int i = 0; i < e; i++){
        cin >> a >> b;
        if (get_parent(parent, a) != get_parent(parent, b))
            union_parent(parent, a, b);
        else {
            cout << i + 1 << endl;
            return 0;
        }
    }
    cout << "0" << endl;
    return 0;
}

File: 백준/Gold/20040. 사이클 게임/사이클 게임.py
import sys


def get_parent(p, x):
    if p[x] != x:
        p[x] = get_parent(p, p[x])
    return p[x]


def union_parent(p, a, b):
    a = get_parent(p, a)
    b = get_parent(p, b)
    if a > b:
        p[a] = b
    elif a < b:
        p[b] = a
    else:
        return


v, e = map(int, sys.stdin.readline().rstrip().split())
parent = [i for i in range(v)]
for i in range(e):
    v1, v2 = map(int, sys.stdin.readline().rstrip().split())
    if get_parent(parent, v1) != get_parent(parent, v2):
        union_parent(parent, v1, v2)
    else:
        print(i + 1)
        exit(0)
print(0)
